rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,221 @@
+
+### Check: detailed features of GOV.UK work
+
+A previous draft of this RFC would have required a detailed audit of the functionality of each app, to ensure we had Smoke or sandboxed E2E tests for each of its features, for some definition of "feature".
+
+This approach was rejected because:
+
+- There is no clear way to demarcate features of an app, let alone GOV.UK as a whole. Given the sheer volume of functionality in our apps, it's not practical to audit or have realistic tests for all of it.
+
+- While we have existing tests for specific features of GOV.UK [[1](https://github.com/alphagov/publishing-e2e-tests/tree/b4a7961cdcd200623a694baf833724a41c98d051/spec/collections_publisher)] [[2](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/csv_preview.feature)], there is no thorough, positive logic to justify their existence or extension [[1](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-050-do-end-to-end-testing-of-gov-uk-applications.md#solution)] [[2](https://github.com/alphagov/publishing-e2e-tests/blob/f12ce7a9626b38728758cf157ea883907bd89de2/docs/what-belongs-in-these-tests.md)]. This is a weak foundation to try and build upon.
+
+## Consequences
+
+### Being more cautious
+
+We will no longer have the safety blanket of our changes languishing in Integration or Staging. It may take some time for us to get used to this new way of working e.g. by gradually doing more [branch deployments](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#review-your-own-changes).
+
+> **Quote from the trial of CD**: "We'll have to up our game to check things before deploying them. There are always corner cases you don't expect. We're perfectly capable of checking all our changes."
+
+### Automatic deployments
+
+We will enable automatic deployments for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145), but only when they meet the safety criteria of this RFC. To help with this, a rough audit of "compatibility" can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+- We can immediately enable automatic deployments for apps marked with a ✅.
+- We will undertake work to address all the issues (⚠) for apps marked with a ❌.
+
+> **We will disable CD for Publishing API**. This was only enabled on a trial basis, and requires non-trivial work to bring it up to the new standard. We should avoid setting a false precedent by leaving CD enabled.
+
+The following steps must be taken as part of enabling automatic deployments:
```

Comment:
I agree we do need more testing and better tools than we have now. 

On the subject of better tooling, I see that as out-of-scope for this RFC: improving the tooling around our Smoke tests should happen naturally if we start to use them more e.g. [here](https://github.com/alphagov/smokey/pull/693).

Flakiness is also reduced in the current CD implementation: we only run the tests that are relevant to an app, so most of the time we're not hindered by flakiness due to the volume of tests, or unrelated apps.

I'd like your help with unpacking the last part, though: how many Smoke tests do we need before we can turn on CD for an app? Originally I thought it would be [a lot](file:///Users/benthorner/govuk/govuk-rfcs/rfc-128-continuous-deployment.md#check-detailed-features-of-govuk-work), but the more I thought about it, the more I convinced myself that the checks [outlined above](file:///Users/benthorner/govuk/govuk-rfcs/rfc-128-continuous-deployment.md#safety-concerns) are enough to give us a manageable level of risk.

Perhaps I'm overthinking your suggestion, though: maybe I just need to really emphasise that all the stated criteria must be met, (which includes at least one new Smokey test for most apps)?

Reply:
> better tooling, I see that as out-of-scope

I agree 👍 

> the checks outlined above are enough to give us a manageable level of risk

I think we should verify a newly deployed app is not just running, but healthy, in a production environment.

Perhaps the line "Check: Smoke tests pass (i.e. the app runs in a production environment)", could be changed to "Check: Health checks pass (i.e. the app is healthy in a production environment)". Does you think that makes sense?

Rather than specifically "smoke tests" I mean "a signal that would tell us if a newly deployed app is healthy", so we can rollback and halt a release.

This signal could be any/all of a Prometheus alert, an app healthcheck endpoint, smoke tests, or something else, provided that this signal can effectively replace an 'eyeball' test for an app.

If we already had that deploy signal, and good rollback behaviour, perhaps we could enable CD today.

> how many Smoke tests do we need before we can turn on CD for an app? 

Fewer smoke tests would be better.

Our current deploy guidance instructs developers to check application metrics in Grafana. The only health signal on app dashboards is usually the 5xx rate and the processes count. We could automate checking an app's 5xx rate (and other relevant metrics) in staging/integration before a release is promoted.

Perhaps we could remove/reduce the smoke tests, if the healthchecks/alerts are sufficient. What do you think?